src/blas_like/level3/Gemm/NN_Multistream.hpp

```diff
@@ -0,0 +1,405 @@
+void SUMMA_NNB_impl_multistream(T alpha,
+                                AbstractDistMatrix<T> const& APre,
+                                AbstractDistMatrix<T> const& BPre,
+                                AbstractDistMatrix<T>& CPre)
+{
+    LogicError("SUMMA_NNB_impl_multistream type-device combo not supported.");
+}
+
+template <typename T, typename=EnableIf<IsDeviceValidType<T,Device::GPU>>>
+void SUMMA_NNC_impl_multistream(
+    T alpha,
+    AbstractDistMatrix<T> const& APre,
+    AbstractDistMatrix<T> const& BPre,
+    AbstractDistMatrix<T>& CPre)
+{
+    EL_DEBUG_CSE;
+
+    constexpr auto D = Device::GPU;
+
+    AUTO_PROFILE_REGION(
+        "SUMMA.NNC.multistream",
+        SyncInfoFromMatrix(
+            static_cast<Matrix<T,D> const&>(CPre.LockedMatrix())));
+
+    const Int sumDim = APre.Width();
+    const Int bsize = Blocksize();
+    const Grid& g = APre.Grid();
+    auto const num_blocks = (sumDim + bsize - 1) / bsize;
+
+    // Setup proxies to assert data layout
+    // (TRB 10/09/2019): Is this really necessary?
+    //   --> For C yes, for A and B, probably not. But having it
+    //       guarantees the description "allgather-allgather-compute".
+    DistMatrixReadProxy<T,T,MC,MR,ELEMENT,D> AProx(APre);
+    DistMatrixReadProxy<T,T,MC,MR,ELEMENT,D> BProx(BPre);
+    DistMatrixReadWriteProxy<T,T,MC,MR,ELEMENT,D> CProx(CPre);
+    auto& A = AProx.GetLocked();
+    auto& B = BProx.GetLocked();
+    auto& C = CProx.Get();
+
+    auto SyncManager = MakeMultiSync(
+        SyncInfoFromMatrix(C.LockedMatrix()),
+        SyncInfoFromMatrix(A.LockedMatrix()),
+        SyncInfoFromMatrix(B.LockedMatrix()));
+
+    // Get the sync pool.
+    auto const& stream_pool = GetSyncInfoPool(C.Grid());
+    auto const num_stream_teams =
+        stream_pool.Size() == 1UL
+        ? 1UL
+        : std::min(stream_pool.Size() / 2, size_t(num_blocks));
+
+    // Setup the temporary matrices. One of each per "stream team".
+    std::vector<DistMatrix<T,MC,STAR,ELEMENT,D>> A1_MC_STAR;
+    std::vector<DistMatrix<T,MR,STAR,ELEMENT,D>> B1Trans_MR_STAR;
+    std::vector<DistMatrix<T,MC,MR,ELEMENT,D>> C_TMP;
+
+    A1_MC_STAR.reserve(num_stream_teams);
+    B1Trans_MR_STAR.reserve(num_stream_teams);
+    C_TMP.reserve(num_stream_teams);
+
+    // Basic setup functions for the temp matrices; also, assign data
+    // to streams in a round-robin fashion.
+    for (auto id = 0UL; id < num_stream_teams; ++id)
+    {
+        auto const& stream_one = stream_pool.Next();
+        auto const& stream_two = stream_pool.Next();
+
+        auto A1 = A1_MC_STAR.emplace(A1_MC_STAR.end(), g);
+        auto B1 = B1Trans_MR_STAR.emplace(
+            B1Trans_MR_STAR.end(), g);
+        auto C1 = C_TMP.emplace(C_TMP.end(), g);
+
+        // A and B are logically const; these just need to have the
+        // right alignment and "stream affinity".
+        SetSyncInfo(A1->Matrix(), stream_one);
+        SetSyncInfo(B1->Matrix(), stream_two);
```

Comment:
`stream_one` should synchronize against `A` and `stream_two` against `B`.

Reply:
Fixed.